src/decrypt.ts

```diff
@@ -1,174 +1,50 @@
-  decipher: DecipherGCM,
+export function startDecryptionStreamWithEmitter(
+  id: JobID,
+  readableStream: ReadableStream,
   contentLength: number | null,
-  id: string | number,
-  key: Buffer,
-  iv: Buffer,
-  authTag: Buffer,
-  ignoreAuthTag = false,
+  key: Uint8Array,
+  iv: Uint8Array,
+  authTag: Uint8Array,
+  ignoreAuthTag?: boolean,
 ): ReadableStream {
-  let totalBytesRead = 0;
-
-  // TransformStreams are supported
-  if ('TransformStream' in self) {
-    return stream.pipeThrough(
-      new (TransformStream as typeof self.TransformStream)({
-        transform: (chunk, controller) => {
-          const bufferChunk = toBuffer(chunk);
-
-          // Decrypt chunk and send it out
-          const decryptedChunk = decipher.update(bufferChunk);
-          controller.enqueue(decryptedChunk);
-
-          // Emit a progress update
-          totalBytesRead += bufferChunk.length;
-          emitProgress('decrypt', totalBytesRead, contentLength, id);
-        },
-        flush: (controller) => {
-          // Finalize decryption when stream is done
-          if (!ignoreAuthTag) {
-            const final = decipher.final();
-            if (final.length > 0) {
-              controller.enqueue(final);
-            }
-          }
-          emitJobCompletion(id, { key, iv, authTag });
-        },
-      }),
-    );
-  }
-
-  let finished = false;
-  // TransformStream not supported, revert to ReadableStream
-  const reader = stream.getReader();
-  return new ReadableStream({
-    /**
-     * Controller
-     * @param controller - Controller
-     */
-    start(controller) {
-      /**
-       * Push on
-       */
-      function push(): void {
-        reader.read().then(({ done, value }) => {
-          if (done) {
-            if (!ignoreAuthTag) {
-              decipher.final();
-            }
-            if (!finished) {
-              controller.close();
-              finished = true;
-            }
-            emitJobCompletion(id, { key, iv, authTag });
-            return;
-          }
-
-          const chunk = toBuffer(value);
-
-          // Decrypt chunk
-          const decValue = decipher.update(chunk);
-
-          // Emit a progress update
-          totalBytesRead += chunk.length;
-          emitProgress('decrypt', totalBytesRead, contentLength, id);
-
-          controller.enqueue(decValue);
-          push();
-        });
-      }
-      push();
-    },
-    /**
-     * On cancel of the decryption stream, throw an error
-     * @param reason - The reason for the cancellation
-     */
-    async cancel(reason) {
-      const err = new PenumbraError(
-        `Decryption stream was cancelled: ${reason}`,
-        id,
-      );
-      logger.error(err);
-      await reader.cancel(err);
-      emitError(err);
-      throw err;
-    },
+  // Construct the decryption stream
+  const decryptionStream = createDecryptionStream(key, iv, {
+    authTag,
+    __dangerouslyIgnoreAuthTag: ignoreAuthTag ?? false,
```

Comment:
I'm not familiar with the underlying property `__dangerouslyIgnoreAuthTag` here, but if it is truly risky to use, would it be worthwhile to rename `ignoreAuthTag` to use the `__dangerously` prefix?

Reply:
But I may do this down the road